js/data/bucket.js

```diff
@@ -337,32 +328,96 @@ function createElementBufferType(components) {
     });
 }
 
-function createPaintAttributes(bucket) {
-    var attributes = {};
-    for (var interfaceName in bucket.programInterfaces) {
-        var layerPaintAttributes = attributes[interfaceName] = {};
-
-        for (var c = 0; c < bucket.childLayers.length; c++) {
-            var childLayer = bucket.childLayers[c];
-            layerPaintAttributes[childLayer.id] = { attributes: [], uniforms: [] };
+function createDataLayers(bucket) {
+    var layers = {};
+
+    for (var dataLayerTypeName in bucket.dataLayerTypes) {
+        var dataLayerType = bucket.dataLayerTypes[dataLayerTypeName];
+        var dataTypeLayers = layers[dataLayerTypeName] = {};
+
+        for (var c = 0; c < bucket.styleLayers.length; c++) {
+            dataTypeLayers[bucket.styleLayers[c].id] = {
+                attributes: [],
+                uniforms: [],
+                defines: [],
+                vertexPragmas: {},
+                fragmentPragmas: {}
+            };
         }
 
-        var interface_ = bucket.programInterfaces[interfaceName];
-        if (!interface_.paintAttributes) continue;
-        for (var i = 0; i < interface_.paintAttributes.length; i++) {
-            var attribute = interface_.paintAttributes[i];
+        if (!dataLayerType.paintAttributes) continue;
+        for (var i = 0; i < dataLayerType.paintAttributes.length; i++) {
+            var attribute = dataLayerType.paintAttributes[i];
+            attribute.multiplier = attribute.multiplier || 1;
+
+            for (var j = 0; j < bucket.styleLayers.length; j++) {
+                var styleLayer = bucket.styleLayers[j];
+                var layer = dataTypeLayers[styleLayer.id];
+
+                var attributeType = attribute.components === 1 ? 'float' : 'vec' + attribute.components;
+                var attributeInputName = attribute.name;
+                assert(attribute.name.slice(0, 2) === 'a_');
+                var attributeInnerName = attribute.name.slice(2);
```

Comment:
Can this slice be removed, either by using the `a_` version in the shaders or by removing `a_` from all the attribute names?